### ISSUE TITLE
Combining pages destroys form inputs

### ISSUE THREAD

**Issue by entrity** [qpdf/qpdf#509]
I've followed the instructions from https://github.com/qpdf/qpdf/issues/11 to combine pdf pages into a single document with the following command:
```bash
qpdf --empty --pages *.pdf -- out.pdf
```
...and in the output pdf, none of my form fields (text inputs & checkboxes) are interactable.

**Comment by jberkenbilt** [comment]
@entrity Are you using qpdf 10.2? I just released it a couple of hours ago, and 10.2 includes a fix for this very issue. Try again with 10.2. It should work. You may nee d to wait a day or two before it shows up in whatever distribution you're using. If Windows, you can download now. If Ubuntu, the ppa is already updated. Otherwise, it may take some time for the versions to show up, or you can grab it and build from source.

**Comment by entrity** [comment]
@jberkenbilt that's great timing. After building 10.2 from source, I still see an issue, though:

Although the form inputs are editable, the text inputs appear to be blank when they aren't highlighted.

In the original pages, I can type into a text input then move on and focus another element, and the text in the first input remains visible; but in the output pdf, after I unfocus a text input, whatever text it contains becomes invisible. If I return focus to the text element, its content becomes visible again (until it loses focus again).

**Comment by jberkenbilt** [comment]
Any way you can share the file with exact steps to reproduce it? It sounds like either the value isn't being properly updated in the form field or the appearance stream is messed up -- these are nuances of how form fields work. It could be a problem with evince, or could be some structural issue introduced by qpdf. If the latter, I will want to find it quickly and get 10.2.1 out with a fix. I have found several bugs in evince doing qpdf development. If you can, try okular. It has more mature support for forms. Alas, there are also things okular messes up that evince does right.

**Comment by entrity** [comment]
@jberkenbilt Thanks.

In okular, when 'Show forms' is not turned on, all of the form inputs look like unfocused inputs.
![image](https://user-images.githubusercontent.com/950824/109075487-0f6b2800-76ae-11eb-9c2e-c7e0d62fe7cd.png)
When I turn on 'Show forms', all of the form inputs show the appearance that I would expect for an in-focus element in evince, Adobe Reader, or a Chromium browser (even though none of them is in focus), and I can interact with them.
![image](https://user-images.githubusercontent.com/950824/109075513-172acc80-76ae-11eb-8b6f-be48385e4984.png)
![image](https://user-images.githubusercontent.com/950824/109075533-201b9e00-76ae-11eb-8c5b-4118f01d0571.png)
After interacting with form inputs and then selecting 'Hide forms', any text entered into a text field disappears, and the field looks just as it does in evince; however, ticked checkboxes successfuly adopt a ticked but 'unfocused' appearance.
![image](https://user-images.githubusercontent.com/950824/109075567-2a3d9c80-76ae-11eb-907a-00ed7d007b36.png)

I've see the term "apperance" used when looking into the issue of unfocused text inputs and checkboxes failing to show their values, as in [this StackOverflow answer](https://stackoverflow.com/a/34611473/507721). I suppose that the page-combining action is causing an appearance is being lost from the PDF dictionary.

Here are the pdfs I used for this example:
I made an extra copy of [a.pdf](https://smallpdf.com/shared#st=5c850e42-654e-4ba6-8970-72ab1d48c845&fn=a.pdf&ct=1614206823617&tl=share-document&rf=link), and then stitched them together with `qpdf --empty --pages a.pdf a-Copy.pdf -- combined.pdf` to make [combined.pdf](https://smallpdf.com/shared#st=d319908c-4abb-43a2-86ec-f5a6a383c33c&fn=combined.pdf&ct=1614206944769&tl=share-document&rf=link)

**Comment by jberkenbilt** [comment]
At a glance, it seems qpdf is putting too much stuff at the top level of the `/AcroForm` dictionary. I'll study it more deeply, but it looks like the brand new code that handles this is not quite right, and I will probably need to release a 10.2.1 with a fix.

**Comment by jberkenbilt** [comment]
Actually, the smallpdf platform is modifying the files. I can tell that combined.pdf is not written by qpdf. Can you just directly attach the files to the github issue?

**Comment by jberkenbilt** [comment]
If it turns out to be a bug in qpdf, I'll give it a few days to see if anything else comes up and then release 10.2.1. I'll say the code to do this is some of the most complex code in qpdf and took me several days to write. There may be a few loose ends.

**Comment by entrity** [comment]
@jberkenbilt thanks for the update.

I've tried attaching the example pdfs to this issue, but:
1. pdf isn't an allowed format
2. tgz isn't an allowed format 
3. Attempts to upload a .zip or a .tar.gz result in "Something went really wrong, and we can't process that file."

Do you want to make a write-only folder in dropbox or google drive where I can drop the files?

**Comment by jberkenbilt** [comment]
Okay, I think I found the problem. It's not what I originally thought. qpdf is copying the correct fields. It looks like it is not getting everything it needs out of the original /AcroForm dictionary. It's probably a small fix, so I'll get it out in 10.2.1 soon. Thanks.

**Comment by jberkenbilt** [comment]
BTW, as a workaround, for now, if you specify either a.pdf or b.pdf as the original, it will work. `qpdf a.pdf --pages a.pdf b.pdf -- combined.pdf`

It's lucky that you happened to bump into this so soon after 10.2 so I can get it fixed before 10.2 spreads to widely. :-)

**Comment by jberkenbilt** [comment]
For what it's worth, it seems that okular/atril/evince also have a bug. Although qpdf failed to copy the `/DR` field from the original /AcroForm dictionary, the PDF file does properly place them on the individual form fields. It looks like the other tools are not looking at it there. The visibility of the filled in field works properly in chrome and in the Preview app on the mac, but those tools have a different problem, which is that the two text fields modify together and the two checkboxes modify together. This is because the two form fields have the same name since they are copies of the same field and qpdf has not done anything to disambiguate them. Those two tools behave identically. I bet Preview on the mac uses whatever chrome uses to display PDF, which would not be surprising.

The file works correctly in Adobe Acrobat Reader.

Form fields are one of the most screwed up bits of the PDF spec. It's no wonder so many tools get it wrong. Hopefully soon qpdf will get it more right than these other tools. :-)

**Comment by jberkenbilt** [comment]
I've studied this a bit more, and I'm quite convinced that this behavior is mostly due to a bug in whatever PDF backend is likely shared between okular, atril, and evince, all of which behave the same way. Those viewers are definitely not properly following the spec regarding resolving resources used by appearance streams. There is a bug in my implementation in that inheritable items are not being copied properly, but even if I fix that, those tools still only render the forms properly if the resources appear at the top level, not when the appear only on the field, and that is definitely not to spec. Working around that would be a substantial amount of work. I am going to fix the bug in qpdf regarding inherited fields and possibly also add some code to disambiguate conflicting field names, and then I'm going to do some digging to figure out how to report the bug to the other tools.

As it happens, the information that qpdf is failing to copy in this particular case is already duplicated at the field level, and so in spite of the qpdf problem, this file should work.

I also have to retract my statement about preview and chrome being wrong. In fact, they are following the spec in that fields with the same name are supposed to share values. So it seems that okular, evince, and atril are getting these wrong. Preview and chrome are following spec.

Interesting, Acrobat Reader does not follow the spec and treats the fields as independent. The full Acrobat disallows editing of the fields on page 2, which have the same name as the fields on page 1. If I change the names of the ones on page 2, then Acrobat allows editing them. Firefox handles it the same as mac preview and Adobe Reader -- it treats the forms as independent and saves independent values for them and generates independent appearance streams, which are then broken once loaded into evince, okular, and atril.

Believe it or not, I don't find any of this particular surprising. When I was working on flattening annotations and handling all the various programs out there, I found a long list of bugs in all sorts of readers. The thing is that qpdf allows you to create PDF files that differ from what lots of tools do, and the PDF spec is so arcane in certain ways. I have often had to figure out what the spec means by just trying things out in lots of different tools. Sadly, for most of the world, the test of whether a PDF file works or not is whether Adobe Acrobat Reader can open it.

So, bottom line: I will fix the bug in qpdf that omits inherited form field data, but when I do, the files will still not behave properly with okular, evince, and atril. I will post a bug report as soon as I figure out where to post it.

**Comment by jberkenbilt** [comment]
I will also add something to the docs explaining this situation.

**Comment by jberkenbilt** [comment]
Okay, blargh, I just reread chapter 12 of the spec. Actually I think maybe okular and friends are following the spec. It looks like the /DR key is not actually inherited. I'm not going to say anything else about this until I'm certain I am reading the spec correctly. It's clear this is not going to be as easy a fix as I initially thought.

**Comment by jberkenbilt** [comment]
I think if I just do the hard thing it will, in the end, work with all of these different applications. It seems that the differences are everyone's attempts to handle erroneous files.

**Comment by jberkenbilt** [comment]
Wow, this turned out to be a lot more complicated than I anticipated, but it is basically working on my work branch. I have a little more error testing to take care of as well as dealing with conflict field names if I determine that's actually needed. Otherwise, it's working pretty well now on these specific example files and a bunch of others I conjured up. It will definitely be 10.3.0 -- I had to add lots of new methods and tweak things to make it work. Definitely one the more complicated bug fixes. Hopefully I get can 10.3.0 out in the next few days.

**Comment by jberkenbilt** [comment]
@entrity -- the attached file is what qpdf 10.3.0 will produce. It appears to function properly across all the viewers I have except some viewers complain about the font, but that's true with the original inputs as well. See if this looks like what you're expecting. I'll try to get 10.3.0 out today or tomorrow if I can.

[combined.pdf](https://github.com/qpdf/qpdf/files/6085563/combined.pdf)

**Comment by jberkenbilt** [comment]
A
[combined.pdf](https://github.com/qpdf/qpdf/files/6085609/combined.pdf)
ctually, this one is a little better.

**Comment by jberkenbilt** [comment]
This is it unless anyone finds any problems. I've finished making code changes for 10.3.

[combined.pdf](https://github.com/qpdf/qpdf/files/6086346/combined.pdf)
